File: src/main/java/algo/interval/DivideIntervalsIntoMinimumNumberOfGroups_2406.java
package algo.interval;

import java.util.*;

public class DivideIntervalsIntoMinimumNumberOfGroups_2406 {
    public static int minGroups(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] == b[0]) {
                return a[1] - b[1];
            }
            return a[0] - b[0];
        });

        // important: sort by interval end
        var pq = new PriorityQueue<int[]>(Comparator.comparingInt(a -> a[1]));
        for (var interval : intervals) {
            if (!pq.isEmpty() && pq.peek()[1] < interval[0]) {
                pq.poll();
            }
            pq.offer(interval);
        }
        return pq.size();
    }

    public static void main(String[] args) {
        System.out.println(minGroups(new int[][]{{5,10},{6,8},{1,5},{2,3},{1,10}}));
        System.out.println(minGroups(new int[][]{{1,3},{5,6},{8,10},{11,13}}));
    }
}

File: src/main/kotlin/algo/interval/DivideIntervalsIntoMinimumNumberOfGroups_2406.kt
package algo.interval

import java.util.PriorityQueue

class KtDivideIntervalsIntoMinimumNumberOfGroups_2406 {
    fun minGroups(intervals: Array<IntArray>): Int {
        val pq = PriorityQueue<Int>()  // last interval of each group sorted by end
        intervals.sortWith(compareBy<IntArray> {it[0]}.thenBy { it[1] })

        intervals.forEach {
            val (start, end) = it
            // no overlap between current interval the last interval so we add current interval to last group
            if (pq.isNotEmpty() && pq.peek() < start) {
                pq.poll()
            }
            pq.offer(end)
        }
        return pq.size
    }
}

fun main() {
    val divide = KtDivideIntervalsIntoMinimumNumberOfGroups_2406()
    println(divide.minGroups(arrayOf(intArrayOf(5,10), intArrayOf(6,8), intArrayOf(1,5), intArrayOf(2,3), intArrayOf(1,10))))
}